src/projectM-sdl/pmSDL.cpp

```diff
@@ -61,27 +61,45 @@ void projectMSDL::audioInputCallbackS16(void *userdata, unsigned char *stream, i
     app->pcm()->addPCM16(pcm16);
 }
 
-SDL_AudioDeviceID projectMSDL::selectAudioInput(int _count) {
-    // TODO: implement some sort of UI allowing the user to select which audio input device they would like to use
-
-
-    // ask the user which capture device to use
-    // printf("Please select which audio input to use:\n");
-    printf("Detected devices:\n");
-    for (int i = 0; i < _count; i++) {
-        printf("  %i: 🎤%s\n", i, SDL_GetAudioDeviceName(i, true));
-    }
-
-    return 0;
-}
-
 int projectMSDL::toggleAudioInput() {
-    
-    CurAudioDevice++;
-    if (CurAudioDevice >= NumAudioDevices)
+    // trigger a toggle with CMD-I or CTRL-I
+    if (wasapi) { // we are currently on WASAPI, so we are going to revert to a microphone/line-in input.
+        if (this->openAudioInput())
+            this->beginAudioCapture();
         CurAudioDevice = 0;
-    selectedAudioDevice = CurAudioDevice;
-    initAudioInput();
+        selectedAudioDevice = CurAudioDevice;
+        this->wasapi = false; // Track wasapi as off so projectMSDL will stop listening to WASAPI loopback in pmSDL_main.
+    }
+    else {
+        this->endAudioCapture(); // end current audio capture.
+        CurAudioDevice++; // iterate device index
+        if (CurAudioDevice >= NumAudioDevices) { // We reached outside the boundaries of available audio devices.
+            CurAudioDevice = 0; // Return to first audio device in the index.
+#ifdef WASAPI_LOOPBACK
```

Comment:
would really love to see this code live in "driver" subclasses rather than all `#ifdef`'d in between the rest of the code in this big file someday

Reply:
I totally agree the #ifdef stuff looks sloppy and confusing when nested like this. I didn't want to undo what someone else did so I kept with it. But hopefully the comments will keep it OK for now. 

I did A LOT of testing so it should work on windows / mac without issue.